File: CreationalPattern/AbstractFactoryPattern/Java/Plants/WallNut.java

package Plants;

import Interface.Plant;

/**
 * Class for Wallnut
 */
public class WallNut implements Plant {
    // Var params for Wallnut
    private String name;
    private int hp;
    private int price;
    
    // Constructor
    public WallNut() {
        this.name = "Wallnut";
        this.hp = 300;
        this.price = 50;
    }

    public WallNut(String name, int hp, int price) {
        this.name = name;
        this.hp = hp;
        this.price = price;
    }
    
    // Override method Display from Plant
    @Override
    public void display() {
        System.out.println("Name: " + name + ", Price: " + price + ", HP: " + hp);
    }

    // Override method GetHealth from Plant
    @Override
    public int getHealth() {
        return hp;
    }

    // Getter and setter
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getPrice() {
        return price;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public void setHp(int hp) {
        this.hp = hp;
    }
}

File: CreationalPattern/AbstractFactoryPattern/Golang/Plant/WallNut.go

package Plant

/**
 * WallNut struct, implement Plant interface
 */
type WallNut struct {
	Name  string
	Price int
	// TODO add more attribute ...
}

// Constructor for WallNut
func NewWallNut(params ...interface{}) *WallNut {
	p := &WallNut{
		Name:  "WallNut",
		Price: 50,
	}

	len := len(params)

	if len > 0 {
		p.Name = params[0].(string)
	}

	if len > 1 {
		p.Price = params[1].(int)
	}

	return p
}

func (s *WallNut) CharacterName() string {
	return s.Name
}

func (s *WallNut) PlantPrice() int {
	return s.Price
}
